datafusion/sql/src/statement.rs

```diff
@@ -710,6 +710,25 @@ impl<S: ContextProvider> SqlToRel<'_, S> {
                     *statement,
                     &mut planner_context,
                 )?;
+
+                //Get inferred data_types from the plan if it is empty in the prepare statement
+                let data_types = match data_types.is_empty() {
+                    true => {
+                        let mut data_types: Vec<DataType> = plan
+                            .get_parameter_types()?
+                            .iter()
+                            .filter_map(|d| match d {
+                                (_, Some(v)) => Some(v.clone()),
+                                _ => None,
+                            })
+                            .collect::<Vec<DataType>>();
+                        data_types.sort();
+                        planner_context.with_prepare_param_data_types(data_types.clone());
+                        data_types
+                    }
+                    false => data_types,
+                };
```

Comment:
If you make `data_types` mutable above, you can do something like this which is a bit more terse:

```suggestion
                if data_types.is_empty() {
                    for dt in plan.get_parameter_types()?.into_values().into_iter() {
                        if let Some(dt) = dt {
                            data_types.push(dt);
                        }
                    }
                    data_types.sort();
                    planner_context.with_prepare_param_data_types(data_types.clone());
                }
```

Reply:
I like it, it looks simpler, thanks!